res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,774 @@
+        }
+    } else {
+        // enable all LEDs
+        for (hotcuePad = 1; hotcuePad <= 8; hotcuePad++) {
+            midi.sendShortMsg(status, 0x40 + hotcuePad-1, 0x7F);
+        }
+    }
+    // shift keyboard Pad 7 and 8 are always enabled
+    midi.sendShortMsg(status+1, 0x46, 0x7F);
+    midi.sendShortMsg(status+1, 0x47, 0x7F);
+};
+
+
+PioneerDDJ400.keyboardModePad = function(channel, control, value, _status, group) {
+    "use strict";
+    channel = (channel & 0xf) < 10 ? 0 : 1;
+    var padNum = (control & 0xf) + 1;
+    var hotcuePad = this.keyboardHotCuePoint[channel];
+
+    // if no hotcue is set for keyboard mode set on first press on a pad
+    if (hotcuePad === 0 && value !== 0) {
+        hotcuePad = padNum;
+        this.keyboardHotCuePoint[channel] = hotcuePad;
+        // if there is no hotcue at this pad, set current play position
+        var hotcuePos = engine.getValue(group, "hotcue_"+hotcuePad+"_position");
+
+        if (hotcuePos < 0) {
+            engine.setValue(group, "hotcue_"+hotcuePad+"_set", 1);
+        }
+
+        this.keyboardModeRefCount[channel] = 0; // reset count
+        this.keyboardModeEnabledOutput(channel, group);
+        return;
+    }
+
+    // if hotcue point is set perform coresponding halftone operation
+    if (value > 0) {
+        // count pressed Pad per deck
+        this.keyboardModeRefCount[channel] += 1;
+        var newValue = this.halftoneToPadMap[padNum-1];
+
+        engine.setValue(group, "pitch", newValue);
+        engine.setValue(group, "hotcue_"+hotcuePad+"_gotoandplay", 1);
+    } else {
+        // decrease the number of active Pads, this should minimize unwanted stops
+        this.keyboardModeRefCount[channel] -= 1;
+        if (this.keyboardModeRefCount[channel] <= 0) {
+            engine.setValue(group, "hotcue_"+hotcuePad+"_gotoandstop", 1);
+            engine.setValue(group, "pitch", 0.0); // reset pitch
+            this.keyboardModeRefCount[channel] = 0; // reset refcount to 0
+        }
+    }
+};
+
+PioneerDDJ400.keyshiftModePad = function(_channel, control, value, _status, group) {
+    "use strict";
+    if (value === 0) {
+        return; // ignore release
+    }
+    engine.setValue(group, "pitch", this.halftoneToPadMap[control & 0xf]);
+};
+
+PioneerDDJ400.samplerModeShiftPadPressed = function(_channel, _control, value, _status, group) {
+    "use strict";
+    if (value === 0) {
+        return; // ignore release
+    }
+
+    var playing = engine.getValue(group, "play");
+    // when playing stop and return to start/cue point
+    if (playing > 0) {
+        engine.setValue(group, "cue_gotoandstop", 1);
+    } else { // load selected track
+        engine.setValue(group, "LoadSelectedTrack", 1);
+    }
+};
+
+
+PioneerDDJ400.shiftPressed = function(channel, _control, value) {
+    "use strict";
+    this.shiftState[channel] = value;
+};
+
+PioneerDDJ400.waveFormRotate = function(_channel, _control, value) {
```

Comment:
I thought this had already been removed??

Reply:
So looking into this a bit, it does appear to be working correctly, and is consistent with the DDJ-400 user manual, which specifies that SHIFT + Rotary Encoder controls waveform zoom.  It also respects the preference at Options -> Preferences -> Waveforms -> Synchronize zoom level across all waveforms.

![image](https://user-images.githubusercontent.com/60218289/77987101-82bdbb80-72de-11ea-9a1e-cb28de776b18.png)

Reply:
I guess it doesn't hurt to leave it in, but is there something more useful this could be mapped to?

Reply:
As an alternative to using shift + `<` and `>` to phrase jump, we could do so here?